### PR TITLE
Feature/pme 480 nextjs implementation of oauth agent

### PR DESCRIPTION
**Pull request by mtrojanowski**

### CODE REVIEW

File: doc/Setup.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+## Build and Run the OAuth Agent
+
+Run these commands from the root folder and the API will then listen on HTTP over port 3000:
+
+```bash
+npm install
+npm dev
+```
+
+Test that the API is contactable by running this command from the root folder:
```
Comment (ju-cu) [inline doc/Setup.md:21]: requires to have `api.example.local` set up. Add a note on that or change to `localhost`

**Comment by ju-cu** [comment]
Missing LICENSE file for Apache 2.0: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository

File: README.md [hunk 2]
```diff
@@ -1,34 +1,51 @@
...
-```bash
-npm run dev
-# or
-yarn dev
-```
+The OAuth Agent acts as a modern `Back End for Front End (BFF)` for Single Page Applications.\
+This implementation demonstrates the standard pattern for SPAs:
 
-Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.
+- Strongest browser security with only `SameSite=strict` cookies
```
Comment (ju-cu) [inline README.md:11]: I advice against the use of superlative as it requires a comparison and discussion (strongest compared to what).
Also, the cookie returned is assigned the attribute `HTTP only`.
Suggest edit:
- Strong browser security with `HTTP only` and `SameSite=strict` cookies

File: README.md [hunk 3]
```diff
@@ -1,34 +1,51 @@
...
+- The OpenID Connect flow uses Authorization Code Flow (PKCE) and a client secret
```
Comment (ju-cu) [inline README.md:12]: PKCE != Authorization Code Flow
Suggest edit:
The OpenID Connect flow uses the Authorization Code Flow with [PKCE](https://curity.io/resources/learn/oauth-pkce/) and a client secret.

that link would be suitable as well: https://curity.io/resources/learn/pkce/

File: README.md [hunk 4]
```diff
@@ -1,34 +1,51 @@
...
-## Deploy on Vercel
+## End-to-End SPA Flow
 
-The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/import?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.
+See the below article for details on how to run the end-to-end solution in a browser:
 
-Check out our [Next.js deployment documentation](https://nextjs.org/docs/deployment) for more details.
+- [SPA Code Example](https://curity.io/resources/learn/token-handler-spa-example/)
+
+The end-to-end solution, by default, uses our Express implementation of the OAuth Agent. If you want to run it with this Next.js implementation,
```
Comment (ju-cu) [inline README.md:42]: link to the implementation: `[Express implementation of the OAuth Agent](https://github.com/curityio/oauth-agent-node-express)`

File: README.md [hunk 5]
```diff
@@ -1,34 +1,51 @@
...
+then a few things would have to be changed manually in the build and deployment scripts.
```
Comment (ju-cu) [inline README.md:43]: add link to build script: `https://github.com/curityio/spa-deployments/blob/main/build.sh#L78`
Provide a `Dockerfile` producing the image `oauthagent-standard:1.0.0` to enable easy integration with the example
Then I believe, there is no need to update any deployment scripts...

**Comment by ju-cu** [comment]
consider changing the deployment path from `<base-url>/api/` to `<base-url>/oauth-agent/` as this is how it looks like with other oauth-agents. As the readme refers to the example deployment we should make integration easy.

File: doc/Setup.md (hunk 1 above)
Comment (mtrojanowski) [reply doc/Setup.md:21]: Yes, I meant to change it to localhost. Must have forgotten...

File: README.md (hunk 5 above)
Comment (mtrojanowski) [reply README.md:43]: You would at least have to change the configuration of the reverse proxy as it has to point to this instance of the OAuth Agent. I will leave it like that for now, but we can think of making a generic script that would be able to spin up an environment with any combination of OAuth Agent + OAuth Proxy.

**Comment by mtrojanowski** [comment]
> Provide custom error pages/responses using JSON and appropriate CORS headers

Any particular place where this doesn't work? All the endpoints should return JSON error responses with proper CORS headers. Or do you mean default 404 responses?

File: README.md (hunk 5 above)
Comment (ju-cu) [reply README.md:72]: In the SPA Code Example, everything is setup locally using docker containers, including the reverse proxy.
If there is a Dockerfile that can be used to create an image to run the nextjs application for the oauth agent within a container, users can simply change the link of the repository of the oauth-agent in the build file and the deploy script will build the image and run the container with the oauth agent.

**Comment by ju-cu** [comment]
> > Provide custom error pages/responses using JSON and appropriate CORS headers
> 
> Any particular place where this doesn't work? All the endpoints should return JSON error responses with proper CORS headers. Or do you mean default 404 responses?

I mean default errors. Basically any error that is not covered by the endpoints (for example 404) should be returned properly.

**Comment by ju-cu** [comment]
> As far as I know, Next.js is pretty opinionated on this one, and we can't change the base path.

Aah, because of the hardcoded API Route. I see. That's a shame :( Well, for the spa-example to work, you then have to change the proxy route to route to `oauth-agent:3001/api`. I guess, that's fine.

**Review by ju-cu — APPROVED** [review]

**Review by ju-cu — APPROVED** [review]

**Comment by mtrojanowski** [comment]
> > > Provide custom error pages/responses using JSON and appropriate CORS headers
> > 
> > 
> > Any particular place where this doesn't work? All the endpoints should return JSON error responses with proper CORS headers. Or do you mean default 404 responses?
> 
> I mean default errors. Basically any error that is not covered by the endpoints (for example 404) should be returned properly.

I tried to find some elegant solution to this but unfortunately, I can't. Interestingly enough, it seems that next.js should return JSON 404 responses on the `/api` route, but for some reason, it doesn't... I could add a "catch all" API route and return a 404 response there, but it feels like a workaround, not a real solution. I will leave that for now as is. Maybe it will get changed in future versions of next.js

**Review by gary-archer — APPROVED** [review]
This looks great and can be merged, since it is much better than previously. I would like to do a more thorough review by writing a code example article. I suspect that CORS is not needed here, and I'm interested in the API paths and general architecture also.

**Pull request merged**